gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/MRJobLauncher.java

```diff
@@ -105,60 +105,125 @@
   private final FileSystem fs;
 
   private volatile Job job;
-  private volatile boolean isCancelled = false;
+  private volatile Path mrJobDir;
 
-  public MRJobLauncher(Properties properties)
+  public MRJobLauncher(Properties properties, Properties jobProps)
       throws Exception {
-    this(properties, new Configuration());
+    this(properties, jobProps, new Configuration());
   }
 
-  public MRJobLauncher(Properties properties, Configuration conf)
+  public MRJobLauncher(Properties properties, Properties jobProps, Configuration conf)
       throws Exception {
-    super(properties);
+    super(properties, jobProps);
     this.conf = conf;
     URI fsUri = URI.create(this.properties.getProperty(ConfigurationKeys.FS_URI_KEY, ConfigurationKeys.LOCAL_FS_URI));
     this.fs = FileSystem.get(fsUri, conf);
   }
 
   @Override
-  public void cancelJob(Properties jobProps)
+  public void cancelJob(JobListener jobListener)
       throws JobException {
-    String jobName = jobProps.getProperty(ConfigurationKeys.JOB_NAME_KEY);
-
-    if (this.isCancelled || !Optional.fromNullable(this.job).isPresent()) {
-      LOG.info(String.format("Job %s has already been cancelled or has not started yet", jobName));
-      return;
-    }
-
     try {
-      // Kill the Hadoop MR if has not completed yet
-      if (!this.job.isComplete()) {
+      if (!this.isCancelled && this.job != null && !this.job.isComplete()) {
         LOG.info("Killing Hadoop MR job " + this.job.getJobID());
         this.job.killJob();
       }
       this.isCancelled = true;
+      LOG.info("Cancelled job " + this.jobProps.getProperty(ConfigurationKeys.JOB_ID_KEY));
+      if (jobListener != null) {
+        jobListener.onJobCancellation(this.jobState);
+      }
     } catch (IOException ioe) {
       LOG.error("Failed to kill the Hadoop MR job " + this.job.getJobID(), ioe);
       throw new JobException("Failed to kill the Hadoop MR job " + this.job.getJobID(), ioe);
+    } finally {
+      cleanUpWorkingDirectory();
```

Comment:
It may be better to do this in `GobblinOutputCommitter`. When canceling a job via Azkaban, the process only has 5 ms to run the `cancel()` method, before Azkaban does a hard kill on the Java process (e.g. kill -9 [pid]). So it's rarely the case where Gobblin gets to do much (if any) cleanup. Calling `this.job.killJob()` make take a while, and `cleanUpWorkingDirectory` may take a while also.

Also, the `GobblinOutputCommitter` actually depends on the presence of the `mrJobDir`. It actually deserializes the `WorkUnit`s from the directory, and tries to cleanup the writer staging directories for each `WorkUnit`. This work used to be done in Gobblin, but was moved to the output committer due to the 5 ms timeout behavior. Since the output committer logic runs in the AM it has more time to do additional cleanup.